.circleci/config.yml

```
@@ -186,6 +186,17 @@ defaults:
       - store_test_results: *store_test_results
       - store_artifacts: *artifacts_test_results
 
+  - test_ubsan_clang: &test_ubsan_clang
+      <<: *test_ubuntu2004_clang
+      steps:
+      - checkout
+      - attach_workspace:
+          at: build
+      - run:
+          <<: *run_soltest
+      - store_test_results: *store_test_results
+      - store_artifacts: *artifacts_test_results
```

Looking at the definition of `test_ubuntu2004_clang`, these steps are not needed. They're already defined in `test_ubuntu2004_clang` and this only overwrites them with the same content. The whole dict could be reduced to just this:

```suggestion
  - test_ubsan_clang: &test_ubsan_clang
      <<: *test_ubuntu2004_clang
```

It's different with `test_asan` above because it's based on `test_ubuntu2004`, which differs slightly - it has `run: *run_soltest_all` instead of `run: *run_soltest`. I wonder if it's intentional though because others all use `run_soltest`.

---

This sequence of steps also looks like something that's repeated often and could be easily extracted into a separate dict. We already dicts named `run_soltest` and `run_soltest_all` (that represent just the single step) but this could be named `run_soltest_steps` and `run_soltest_all_steps`.

---

I have addressed the first comment. I don't fully understand the second comment, so leaving it unaddressed until further clarification.

---

I meant that this:
```yaml
      steps:
      - checkout
      - attach_workspace:
          at: build
      - run:
          <<: *run_soltest
      - store_test_results: *store_test_results
      - store_artifacts: *artifacts_test_results
```

is used in several places, with just a slight variation in the `run` part. I.e. sometimes it's `run: *run_soltest` (which means the same thing), and sometimes it's modified to run `run_soltest_all` instead.

You could add an entry for this under the `defaults:` as we already do with some other reusable pieces of config:
```yaml
  - run_soltest_steps: &run_soltest_steps
      steps:
      - checkout
      - attach_workspace:
          at: build
      - run: *run_soltest
      - store_test_results: *store_test_results
      - store_artifacts: *artifacts_test_results
```

And then some other dicts could be simplified. For example `test_ubuntu2004_clang` could be shortened:
```yaml
  - test_ubuntu2004_clang: &test_ubuntu2004_clang
      docker:
        - image: << pipeline.parameters.ubuntu-2004-clang-docker-image >>
      <<: *run_soltest_steps
```
Same with `run_soltest_all_steps`.

I'd be inclined to call them just `run_soltest`/`run_soltest_all` but we already have dicts that use these names (in fact that's what `run: *run_soltest` is referring to):
```yaml
  - run_soltest: &run_soltest
      name: soltest
      no_output_timeout: 30m
      command: ./.circleci/soltest.sh

  - run_soltest_all: &run_soltest_all
      name: soltest_all
      no_output_timeout: 30m
      command: ./.circleci/soltest_all.sh
```